update branching section (#41)
rossyndicate/ROSSyndicate_best_practices: 05-Git-Feature-Branches.Rmd

Author: This is the only file that needs review - others are byproducts of the bookdown render.

## How to Create and Work with Feature Branches

### Step 1: List Existing Branches and Remotes

Before creating a branch, check what already exists:

``` bash
git branch -v          # list local branches
git remote -v          # list existing remote connection (i.e., git repos you can pull files in from)
```

### Step 2: Add a Remote (if needed)

If the remote you want to branch from isn't already listed (e.g., for a PR
review of someone else's fork):

-   Go to the GitHub page and copy the SSH or HTTPS link.
-   Add the remote (because you want to branch to it):

``` bash
git remote add new_remote_name ssh_link
# or
git remote add new_remote_name https_link
```

-   Fetch the remote to see its contents (does not download files yet):

``` bash
git fetch new_remote_name
```

To remove a remote you no longer need:

``` bash
git remote delete remote_name
```

### Step 3: Create or Switch to a Branch

**Create a new branch from a remote branch:**

``` bash
git checkout -b new_branch_name remote_name/branch_name
```

**Switch to an existing branch:**

``` bash
git checkout branch_name
# or
git switch branch_name
```

**Create a new branch starting from the current branch:**

``` bash

> git checkout -b new-branch-name

Reviewer: ```suggestion
git checkout -b new_branch_name
```

```

The branch starts as a copy of whichever branch you were on. If you switch back
to main using `git checkout main`, your branch work will appear to disappear --

> it's still there on the feature branch.

Reviewer: ```suggestion
The branch starts as a copy of whichever branch you were on, unless you specify
a branch to originate from, like shown above. If you switch back
to main using `git checkout main`, your branch work will appear to disappear --
fear not, it's still there on the feature branch.
```

### Step 4: Make and Commit Changes

Add and commit as usual within your branch:

``` bash
git add .
git commit -m "some changes to this branch"
```

### Step 5: Push or Merge Your Branch

**Push the branch to your remote (forked repo):**

``` bash

> git push -u origin new-branch-name

Reviewer: ```suggestion
git push -u origin new_branch_name
```

```

**Merge branch changes back into main:**

``` bash
git checkout main

> git merge new-branch-name

Reviewer: ```suggestion
git merge new_branch_name
```

Sometimes you want to make a PR that starts from where the central repository
(upstream) is, rather than your full main branch. This is useful when you have
a lot of local changes but want to submit only specific files.

``` bash
# Create new branch "pr" from upstream/main
git checkout -b pr upstream/main

# Then selectively add files from your main branch:
git checkout main .gitignore _targets
# or
git restore --source main -- file_name
```

--------------------------------------------------------------------------------

## Merging Branches

Merge combines two branch histories, keeping the most recent commit from each
file and joining the two timelines into one.

``` bash
git merge branch_name
# example:
git merge experiment
```

> **Tip:** If you get stuck in a merge review in the terminal with no obvious
> way to exit, type `:q` and press Enter.

--------------------------------------------------------------------------------

## Moving Work from a Feature Branch to Main

Say you worked on an "experiment" branch and want to bring some or all of that
work into your main branch:

1.  Make sure all changes are committed on your current branch. 2. Create (or

>     switch to) the experimental branch: `bash     git checkout -b experiment`

Reviewer: ```suggestion
1.  Make sure all changes are committed on your current branch. 2. Create (or
    switch to) the experimental branch: 
    ``` bash     
    git checkout -b experiment
    ```
```

2.  Make changes and commit:

>     `bash     git add .     git commit -m "some changes to this branch"`

Reviewer: ```suggestion
2.  Make changes and commit:
    ```bash
    git add .     
    git commit -m "some changes to this branch"
    ```
```

3.  Bring changes into main:
    -   **Update just one file in main:**
        `bash     git switch main     git checkout experiment file_to_checkout`
    -   **Merge all changes from "experiment" into current branch:**

>         `bash     git merge experiment`

Reviewer: ```suggestion
3.  Bring changes into main:
    -   **Update just one file in main:**
        ``` bash     
        git switch main     
        git checkout experiment file_to_checkout
        ```
    -   **Merge all changes from "experiment" into current branch:**
        ``` bash     
        git merge experiment
        ```
```

--------------------------------------------------------------------------------

## `git checkout` vs. `git merge`

checkout: Use to bring in a file from another branch. If the file already
exists, it overwrites local changes with the new checked out version. It does
not combine histories.

merge: Merge can be used to combine two branches or individual files from two
branches. It will combine with rules for keeping most recent commit from each
file, joining the two histories into one.

| | `git checkout` | `git merge` | |---|---|---| | **Use for** | Bringing in a
specific file from another branch | Combining two full branches | | **If file
exists** | Overwrites local changes with the checked-out version | Combines
with rules for keeping the most recent commit | | **History** | Does not

> combine histories | Joins two histories into one |

Reviewer: ```suggestion
| | `git checkout` | `git merge` | 
|---|---|---| 
| **Use for** | Bringing in a specific file from another branch | Combining two full branches | 
| **If file exists** | Overwrites local changes with the checked-out version | Combines with rules for keeping the most recent commit |
 | **History** | Does not combine histories | Joins two histories into one |
```

--------------------------------------------------------------------------------

## Deleting a Branch

Once a branch is no longer needed:

``` bash
git branch -D branch_name
```

> 

Reviewer: ```suggestion
Note, this permanently deletes the branch. You CANNOT retrieve it from your local device. If and only if you pushed the most recent changes to origin will you be able to restore any changes from this branch.
```

--------------------------------------------------------------------------------

## Feature Branch Best Practices

> -   Keep your fork up to date regularly. **COMMUNICATE** with your team.

Reviewer: ```suggestion
-   **COMMUNICATE** with your team.
-   Keep your fork up to date regularly. 
```
